File: sentimental-credit/credit.py
# TODO
from cs50 import get_int, get_string

number = get_int("Number: ")

list_1, list_2 = [], []
n, m = 100, 10
sum_1, sum_2 = 0, 0

for i in range(8):      # 16-digit -> max 8 time
    digit_1 = int((number % n) / (n / 10))
    digit_1 *= 2
    list_1.append(digit_1)
    n *= 100

    if digit_1 < 10:
        sum_1 += digit_1
    else:
        sum_1 += (1 + digit_1 % 10)


for i in range(8):
    digit_2 = int((number % m) / (m / 10))
    list_2.append(digit_2)
    m *= 100

    sum_2 += digit_2

# print(list_1)
# print(list_2)
# print(sum_1)
# print(sum_2)

if ((sum_1 + sum_2) % 10 == 0) and ((int(number % 10000000000000000 / 10000000000000) == 34 or\
                                    int(number % 10000000000000000 / 10000000000000) == 37)):

    print("AMEX\n")

elif ((sum_1 + sum_2) % 10 == 0) and ((int(number % 100000000000000000 / 100000000000000) > 50 and\
                                       int(number % 100000000000000000 / 100000000000000) < 56)):

    print("MASTERCARD\n")

elif ((sum_1 + sum_2) % 10 == 0) and ((int(number % 100000000000000000 / 1000000000000000) == 4 or\
                                       int(number % 100000000000000 / 1000000000000) == 4)):

    print("VISA\n")

else:
    print("INVALID\n")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

// call functions
long card_number(void);
bool is_valid(long number);

// main program
int main(void)
{
    // input card number
    long number = card_number();

    // calculate checksum, return boolean (T/F)
    bool valid = is_valid(number);

    // AMEX - begin with 34 or 37 - 15 digits
    if ((valid && (number % 1000000000000000) / 100000000000000 == 3) && ((number % 100000000000000) / 10000000000000 == 4
            || (number % 100000000000000) / 10000000000000 == 7))
    {
        printf("AMEX\n");
    }

    // MASTERCARD - begin with 51, 52, 53, 54, 55 - 16 digits
    else if ((valid && (number % 10000000000000000) / 1000000000000000 == 5) && ((number % 1000000000000000) / 100000000000000 == 1
             || (number % 1000000000000000) / 100000000000000 == 2 || (number % 1000000000000000) / 100000000000000 == 3
             || (number % 1000000000000000) / 100000000000000 == 4 || (number % 1000000000000000) / 100000000000000 == 5))
    {
        printf("MASTERCARD\n");
    }

    // VISA - begin with 4 - 13/16 digits
    else if (valid && ((((number % 10000000000000) / 1000000000000) == 4 && ((number % 100000000000000) / 10000000000000) == 0)
                       || ((number % 10000000000000000) / 1000000000000000) == 4))
    {
        printf("VISA\n");
    }

    // INVALID - None
    else
    {
        printf("INVALID\n");
    }
}

// input user card number
long card_number(void)
{
    long number = get_long("Number: ");
    return number;
}

// checksum function
bool is_valid(long number)
{
    long mod1 = 100; // second to the last digit
    long div1 = 10;

    long mod2 = 10; // last digit
    long div2 = 1;

    long digit1;
    int digit1_1; // first digit after *2
    int digit1_2; // second digit after *2

    long digit2;

    int ind = 0; // initialize index
    int sum = 0; // initialize sum



    while (ind < 8) // 8 iteration max for all types of card
    {
        digit1 = number % mod1;
        digit1 /= div1;
        digit1 *= 2;
        digit1_1 = digit1 % 10;
        digit1_2 = digit1 / 10;

        digit2 = number % mod2;
        digit2 /= div2;

        sum += digit1_1 + digit1_2 + digit2;
        // printf("sum: %i\n", sum);

        // move digit cursor for each new iteration
        mod1 *= 100;
        div1 *= 100;
        mod2 *= 100;
        div2 *= 100;

        ind ++;
    }

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}